### ISSUE TITLE
[feature - arango] Change all backend mutation implementations to return only ID 

### ISSUE THREAD

**Issue by pxp928** [guacsec/guac#1197]
**Is your feature request related to a problem? Please describe.**
Followup to issue https://github.com/guacsec/guac/issues/1116. Now that the graphQL mutation  API has been updated, the [backend](https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/backends.go) needs to be updated to only return the IDs based on the revised [resolvers](https://github.com/guacsec/guac/tree/main/pkg/assembler/graphql/resolvers)

**Describe the solution you'd like**

Multiple issues will be created for each backend update. This issue will focus on the updates to the [arango](https://github.com/guacsec/guac/tree/main/pkg/assembler/backends/arangodb) backend. 

All mutations must be changed to return only ID (string). 

For example for artifacts:

```
IngestArtifact(ctx context.Context, artifact *model.ArtifactInputSpec) (string, error)
IngestArtifacts(ctx context.Context, artifacts []*model.ArtifactInputSpec) ([]string, error) 
```

Must both be updated in the [backend interface](https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/backends.go#L54-L55) and in the [inmem backend](https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/arangodb/artifact.go#L178-L227).

Optimization to replace `UPSERT` with `INSERT` must also be done.

Unit tests must be updated as needed to reflect these changes

**Comment by arorasoham9** [comment]
Taking this issue.

**Comment by pxp928** [comment]
> If I change the return type in the [backend interface](https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/backends.go#L54-L55) then it creates errors in the ent and neo4j backends. Is this normal behaviour? 

Yes, all the backends will have to be changed simultaneously so that we don't break the main branch. Don't worry about the others backends for your PR.

**Comment by arorasoham9** [comment]
@pxp928 I did make some changes to fix this issue. However, I am getting the following error.

`cannot use arangoClient (variable of type *arangoClient) as backends.Backend value in return statement: *arangoClient does not implement backends.Backend (wrong type for method IngestCertifyLegal)
		have IngestCertifyLegal(context.Context, model.PackageOrSourceInput, []*model.LicenseInputSpec, []*model.LicenseInputSpec, *model.CertifyLegalInputSpec) (*model.CertifyLegal, error)
		want IngestCertifyLegal(context.Context, model.PackageOrSourceInput, []*model.LicenseInputSpec, []*model.LicenseInputSpec, *model.CertifyLegalInputSpec) (string, error)compiler[InvalidIfaceAssign](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#InvalidIfaceAssign)`

I could really use some help. Thank you!

**Comment by pxp928** [comment]
@arorasoham9 You are working on the `inmem` backend correct?

**Comment by arorasoham9** [comment]
For this issue, I was working on the arango backend. However, I think I'll need to fix all backends if we don't want to break main.

**Comment by pxp928** [comment]
@arorasoham9  So maybe for now you can do the fixes that are needed for the various backends and we can optimize in later PRs as needed.

**Comment by pxp928** [comment]
On second thought @arorasoham9, it might be better to start with the `inmem` backend and create a PR and we do not merge until we have PRs for all the backends that we can merge together to not break main.

**Comment by mihaimaruseac** [comment]
I think that might be harder as you'll have to keep the entire set of PRs in sync.

Why not update things by going over each GraphQL entry-point at once, across all backends? So we first make one PR that handles the `ingestArtifact`,`ingestArtifacts` across all backends, then another PR for the next set of mutations and so on? This way, each PR can be merged as soon as it's done and the functionality and API is not broken at all.

**Comment by pxp928** [comment]
the problem with that approach is we need to optimize some of the backends (like arango and perhaps ENT also) to only return IDs (remove upserts where possible). So need some understanding of the backends to make the proper changes

**Comment by mihaimaruseac** [comment]
That could be done as a separate step after the returning ID is pushed down to each backend. Then it can be done for each backend in its own PR.

Right now because the backends are implemented as a Go interface, you need to change the interface function to make it return an ID and that forces you to make all backend implementations to migrate to return an ID. We can use the existing code for each but just return the ID from the struct and then on a separate pass (which could be done in parallel if the database query optimization is localized) we can optimize the DB query

**Comment by pxp928** [comment]
other thing is we could just make the updates to the backends and optimization can come in secondary PRs (as needed)

**Comment by pxp928** [comment]
In that case, if should be fairly straight forward to update the backends to return just IDs with the existing code. Some of the unit tests might have to be updated.

**Comment by pxp928** [comment]
@arorasoham9 So like you were saying before, we need to change all the backend and we can optimize at a later date. @ivanvanderbyl has already written what needs to be done on the ENT side for these changes: https://github.com/guacsec/guac/issues/1198

**Comment by arorasoham9** [comment]
@pxp928 Okay, I'll work on fixing arangoDB and neo4j backends then.

**Comment by pxp928** [comment]
@arorasoham9 just an update here, we made a [branch](https://github.com/guacsec/guac/tree/1116-return-ID) that we can push the PRs to so we do not break main. 

Also, I like @mrizzi suggestion to update the `backend` interface to double up each ingestion (that needs updating) to return the original struct and a new one just to return `string` (for the ID). This way we do not break the other backends and test each backend properly before merging to main.

For example:

```
IngestArtifact(ctx context.Context, artifact *model.ArtifactInputSpec) (*model.Artifact, error)
IngestArtifactID(ctx context.Context, artifact *model.ArtifactInputSpec) (string, error)
```

Once its all done, we can delete the `IngestArtifactID` and refactor the original (`IngestArtifact`) to return just a `string` for the ID.

**Comment by mrizzi** [comment]
FYI https://github.com/guacsec/guac/pull/1285 doubles up each (involved) ingestion

**Comment by arorasoham9** [comment]
@pxp928  Sounds good! I think I understand what you want here. So we are updating all backends together and using the new ingestion instead(for now until we verify everything works correctly)?

**Comment by pxp928** [comment]
yes keep both versions and work on returning the ID on the new ingestion. Once completed, we can remove the old version and just keep the ones that return IDs (by refactoring and keeping the original function name for example: `IngestArtifact`)

**Comment by pxp928** [comment]
https://github.com/guacsec/guac/pull/1529 will complete and close this issues
